dbms/src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -4969,6 +4975,185 @@ void StorageReplicatedMergeTree::replacePartitionFrom(const StoragePtr & source_
+        entry_replace.from_database = src_data.database_name;
+        entry_replace.from_table = src_data.table_name;
+        for (const auto & part : src_parts)
+            entry_replace.src_part_names.emplace_back(part->name);
+        for (const auto & part : dst_parts)
+            entry_replace.new_part_names.emplace_back(part->name);
+        for (const String & checksum : part_checksums)
+            entry_replace.part_names_checksums.emplace_back(checksum);
+        entry_replace.columns_version = columns_version;
+    }
+
+    queue.removePartProducingOpsInRange(zookeeper, drop_range, entry);
+
+    clearBlocksInPartition(*zookeeper, drop_range.partition_id, drop_range.max_block, drop_range.max_block);
+
+    DataPartsVector parts_to_remove;
+    Coordination::Responses op_results;
+
+    try
+    {
+        Coordination::Requests ops;
+        for (size_t i = 0; i < dst_parts.size(); ++i)
+        {
+            dest_table_storage->getCommitPartOps(ops, dst_parts[i], block_id_paths[i]);
+            ephemeral_locks[i].getUnlockOps(ops);
+
+            if (ops.size() > zkutil::MULTI_BATCH_SIZE)
+            {
+                zookeeper->multi(ops);
+                ops.clear();
+            }
+        }
+
+        ops.emplace_back(zkutil::makeCreateRequest(dest_table_storage->zookeeper_path + "/log/log-", entry.toString(), zkutil::CreateMode::PersistentSequential));
+
+        {
+            Transaction transaction(*dest_table_storage);
+
+            auto src_data_parts_lock = lockParts();
+            auto dest_data_parts_lock = dest_table_storage->lockParts();
+
+            for (MutableDataPartPtr & part : dst_parts)
+                dest_table_storage->renameTempPartAndReplace(part, nullptr, &transaction, dest_data_parts_lock);
+
+            op_results = zookeeper->multi(ops);
+
+            parts_to_remove = removePartsInRangeFromWorkingSet(drop_range, true, false, dest_data_parts_lock);
+            transaction.commit(&dest_data_parts_lock);
+        }
+
+        PartLog::addNewParts(global_context, dst_parts, watch.elapsed());
+    }
+    catch (...)
+    {
+        PartLog::addNewParts(global_context, dst_parts, watch.elapsed(), ExecutionStatus::fromCurrentException());
+        throw;
+    }
+
+    String log_znode_path = dynamic_cast<const Coordination::CreateResponse &>(*op_results.back()).path_created;
+    entry.znode_name = log_znode_path.substr(log_znode_path.find_last_of('/') + 1);
+
+    for (auto & lock : ephemeral_locks)
+        lock.assumeUnlocked();
+
+    tryRemovePartsFromZooKeeperWithRetries(parts_to_remove);
+
+    parts_to_remove.clear();
+    cleanup_thread.wakeup();
+
+    if (context.getSettingsRef().replication_alter_partitions_sync > 1)
+        dest_table_storage->waitForAllReplicasToProcessLogEntry(entry);
+
+    Coordination::Requests ops_dest;
+
+    ops_dest.emplace_back(zkutil::makeCreateRequest(zookeeper_path + "/log/log-", entry_delete.toString(), zkutil::CreateMode::PersistentSequential));
```

Comment:
Is there any reason for applying `entry` and `entry_delete` separately? (As well as calling `waitForAllReplicasToProcessLogEntry()` twice)

Reply:
It was applied both because we apply change on two tables, if we apply only one, Zookeeper will apply changes in only one table for all replica